state/application_test.go

```diff
@@ -6054,50 +6147,85 @@ func (s *ApplicationSuite) TestUpdateStorageConstraints(c *gc.C) {
 	err = app.UpdateStorageConstraints(newSC)
 	c.Assert(err, jc.ErrorIsNil)
 
+	// Make sure storage constraints are updated to newSC.
 	cons, err = app.StorageConstraints()
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(cons, gc.HasLen, 1)
-	c.Assert(cons["database"].Pool, gc.Equals, "loop")
-	c.Assert(cons["database"].Count, gc.Equals, uint64(1))
-	c.Assert(cons["database"].Size, gc.Equals, uint64(4096))
+	c.Assert(cons["data"].Pool, gc.Equals, "loop")
+	c.Assert(cons["data"].Count, gc.Equals, uint64(5))
+	c.Assert(cons["data"].Size, gc.Equals, uint64(4096))
 }
 
-func (s *ApplicationSuite) TestUpdateStorageConstraintsNotSupported(c *gc.C) {
+func (s *ApplicationSuite) TestUpdateStorageConstraints_Concurrent_ErrOnCountIncrease(c *gc.C) {
 	oldSC := map[string]state.StorageConstraints{
-		"database": {
+		"data": {
 			Pool:  "loop",
 			Size:  100,
-			Count: 1,
+			Count: 5,
 		},
 	}
-	charm := s.AddTestingCharm(c, "cockroachdb")
-	app := s.AddTestingApplicationWithStorage(c, "cockroachdb", charm, oldSC)
-
+	// storage-constraint1 charm storage constraint size is 1-5. We set this as our original application charm.
+	charm := s.AddTestingCharm(c, "storage-constraint1")
+	app := s.AddTestingApplicationWithStorage(c, "storage-constraint1", charm, oldSC)
 	cons, err := app.StorageConstraints()
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(cons, gc.HasLen, 1)
-	c.Assert(cons["database"].Pool, gc.Equals, "loop")
-	c.Assert(cons["database"].Count, gc.DeepEquals, uint64(1))
-	c.Assert(cons["database"].Size, gc.Equals, uint64(100))
+	c.Assert(cons["data"].Pool, gc.Equals, "loop")
+	c.Assert(cons["data"].Count, gc.DeepEquals, uint64(5))
+	c.Assert(cons["data"].Size, gc.Equals, uint64(100))
 
+	// We try to update original application charm storage constraint size to 3.
 	newSC := map[string]state.StorageConstraints{
-		"database":       makeStorageCons("loop", 4096, 1),
-		"not-supported1": makeStorageCons("loop", 4096, 1),
-		"not-supported2": makeStorageCons("loop", 4096, 1),
+		"data": makeStorageCons("loop", 4096, 3),
 	}
 
+	// storage-constraint2 charm storage constraint size is 6-10.
+	upgradedCharm := s.Factory.MakeCharm(c, &factory.CharmParams{Name: "storage-constraint2", URL: "local:quantal/quantal-storageconstraint1-2"})
+
+	// Set charm to new storage-constraint2 charm with storage constraint size
+	// from 6-10 before running replaceStorageConstraintsOp txn.
+	defer state.SetBeforeHooks(c, s.State, func() {
+		err := app.SetCharm(state.SetCharmConfig{
+			Charm:       upgradedCharm,
+			CharmOrigin: defaultCharmOrigin(upgradedCharm.URL()),
+			StorageConstraints: map[string]state.StorageConstraints{
+				"data": {
+					Pool:  "loop",
+					Size:  200,
+					Count: 6,
+				},
+			},
+		})
+		c.Assert(err, jc.ErrorIsNil)
+	}).Check()
+
+	// This should fail since we try to set storage constraint size to 3 in
+	// the new charm with storage constraint size of range 6-10.
 	err = app.UpdateStorageConstraints(newSC)
 	c.Assert(err, gc.NotNil)
```

Comment:
If there a typed error? eg `errors.Invalid`. We should look to use `c.Assert(err, jc.ErrorIs, errors.Invalid)` for example

Reply:
the existing validation method i.e "validateStorageConstraints" doesn't return a typed error, shall I add that in?

Reply:
If the existing code doesn't have it, we can leave it for now and fix later if we need to

Reply:
No, but we should have used `c.Assert(err, jc.ErrorMatches, "test error1")`.

Reply:
Sure will refactor this in the next PR, thanks